java-checks/src/main/java/org/sonar/java/checks/CastArithmeticOperandCheck.java

```diff
@@ -19,75 +19,124 @@
+    super.visitVariable(tree);
+  }
+
+  @Override
+  public void visitMethodInvocation(MethodInvocationTree tree) {
+    checkMethodInvocationArgument(tree);
+    super.visitMethodInvocation(tree);
+  }
+
+  @Override
+  public void visitMethod(MethodTree tree) {
+    if (tree.is(Tree.Kind.METHOD)) {
+      checkMethodTree((MethodTreeImpl) tree);
+    }
+    super.visitMethod(tree);
+  }
+
+  @Override
+  public void visitBinaryExpression(BinaryExpressionTree tree) {
+    boolean continueVisit = true;
+    if (tree.is(Tree.Kind.DIVIDE) && isIntOrLong(tree.symbolType())) {
+      continueVisit = checkIntegerDivisionInsideFloatingPointExpression(tree);
+    }
+    if (continueVisit) {
+      super.visitBinaryExpression(tree);
+    }
+  }
+
   public void visitNode(Tree tree) {
-    if (hasSemantic()) {
-      Type varType;
-      ExpressionTree expr;
-      switch (tree.kind()) {
-        case ASSIGNMENT:
-          AssignmentExpressionTree aet = (AssignmentExpressionTree) tree;
-          varType = aet.symbolType();
-          expr = aet.expression();
-          checkExpression(varType, expr);
-          break;
-        case VARIABLE:
-          VariableTree variableTree = (VariableTree) tree;
-          varType = variableTree.type().symbolType();
-          expr = variableTree.initializer();
-          checkExpression(varType, expr);
-          break;
-        case METHOD_INVOCATION:
-          checkMethodInvocationArgument((MethodInvocationTree) tree);
-          break;
-        case METHOD:
-          checkMethodTree((MethodTreeImpl) tree);
-          break;
-        case DIVIDE:
-          BinaryExpressionTree binaryExpr = (BinaryExpressionTree) tree;
-          if (isIntOrLong(binaryExpr.symbolType())) {
-            checkIntegerDivisionInsideFloatingPointExpression(binaryExpr);
-          }
-          break;
-        default:
-          throw new IllegalArgumentException("Tree " + tree.kind() + " not handled.");
-      }
+    if (context.getSemanticModel() == null) {
+      return;
+    }
+    Type varType;
+    ExpressionTree expr;
+    switch (tree.kind()) {
+      case ASSIGNMENT:
+        AssignmentExpressionTree aet = (AssignmentExpressionTree) tree;
+        varType = aet.symbolType();
+        expr = aet.expression();
+        checkExpression(varType, expr);
+        break;
+      case VARIABLE:
+        VariableTree variableTree = (VariableTree) tree;
+        varType = variableTree.type().symbolType();
+        expr = variableTree.initializer();
+        checkExpression(varType, expr);
+        break;
+      case METHOD_INVOCATION:
+        checkMethodInvocationArgument((MethodInvocationTree) tree);
+        break;
+      case METHOD:
+        checkMethodTree((MethodTreeImpl) tree);
+        break;
+      case DIVIDE:
+        BinaryExpressionTree binaryExpr = (BinaryExpressionTree) tree;
+        System.out.println("-->"+binaryExpr.operatorToken().line()+" "+binaryExpr.operatorToken().column());
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Replace this use of System.out or System.err by a logger. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS106)